### ISSUE TITLE
## Summary

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5643]
## Summary

Add the **Prodigy** optimizer (Mishchenko & Defazio 2023, arXiv:2306.06101) to
`odyssey.training.optimizers` as a pure-functional step, mirroring the existing
first-order-moment family (`adan`, `adopt`). Prodigy is a D-Adaptation-family
learner that estimates the distance-to-solution `d_t ≈ ‖x₀ − x*‖` online and
uses it to scale an Adam-style update — it has **no learning rate to tune**.

## Motivation

Odyssey ships SGD/Adam/AdamW/RMSprop/LARS and several modern optimizers
(ADOPT, Adan, Sophia, Muon variants, Shampoo, SOAP) but not a parameter-free /
D-Adaptation optimizer. Prodigy fills the "no-LR" slot in the optimizer
taxonomy: it replaces the learning-rate axis with an online distance estimate,
which is useful for tune-nothing training and for LR-ladder-free ablations.

## Algorithm (Adam variant — Algorithm 4 of the paper)

Per step, using the CURRENT `d` in both the accumulators and the parameter
update:

```
m'    = beta1 * m + (1 - beta1) * d * g
v'    = beta2 * v + (1 - beta2) * d^2 * g^2
r'    = sqrt(beta2) * r + (1 - sqrt(beta2)) * gamma * d^2 * <g, x0 - x>
s'    = sqrt(beta2) * s + (1 - sqrt(beta2)) * gamma * d^2 * g
d_hat = r' / ||s'||_1
d'    = max(d, d_hat)                 (monotone non-decreasing)
x'    = x - gamma * d * m' / (sqrt(v') + d * eps)
```

`gamma` is the base step-size schedule (default 1.0). A `growth_rate` cap
limits per-step growth of `d` (natural warmup); the default disables it.

## Scope

- New file `src/odyssey/training/optimizers/prodigy.mojo`:
  `prodigy_step` (+ `prodigy_step_simple` convenience wrapper).
- Pure-functional: all state (per-coordinate `m`/`v`/`s`, scalars `r`/`d`, and
  the `x0` reference) travels in and out — the sibling idiom.
- Export line in `optimizers/__init__.mojo`; README primitive-table row;
  `scripts/run_primitive_test.sh` registration.
- Unit test `tests/odyssey/training/optimizers/test_prodigy.mojo` with a
  committed NumPy parity reference under `parity_refs/`.

## Dtype support

float32 and float64 (parity test in float64; a float32 d-growth test guards the
scalar-reduction dtype read). `params`/`gradients` must share dtype+shape (the
step raises otherwise). Mixed precision is not supported.

## Acceptance criteria

- [x] `prodigy_step` implements Algorithm 4 exactly; d is monotone non-decreasing.
- [x] Parity test over 5 steps asserts params AND the scalar `d` (d grows
      1e-6 → ~3.45e-5) against a committed, re-runnable reference.
- [x] Shape/dtype guards raise (raises-tests).
- [x] `growth_rate` cap test; `step_simple` == defaults test; descent test.
- [x] Registered in `__init__.mojo`, README table, `run_primitive_test.sh`.
